File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3016_minimum-number-of-pushes-to-type-word-ii/solutions/c/solution.c
#include <string.h>
#include <stdlib.h>

int minimumPushes(char* word) {
    int freq[26] = {0};
    int len = strlen(word);

    // Calculate frequency of each letter
    for (int i = 0; i < len; i++) {
        freq[word[i] - 'a']++;
    }

    // Sort frequencies in descending order
    qsort(freq, 26, sizeof(int), (int (*)(const void*, const void*)) strcmp);

    int pushes = 0;
    int k = 1; // Key presses for the position on the key

    // Distribute the most frequent letters to the least number of key presses
    for (int i = 25; i >= 0 && freq[i] > 0; i--) {
        pushes += freq[i] * k;
        if (i % 3 == 0) k++; // Every three characters, increment the key presses needed
    }

    return pushes;
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3016_minimum-number-of-pushes-to-type-word-ii/solutions/python3/solution.py
class Solution:
    def minimumPushes(self, word: str) -> int:
        from collections import Counter
        
        # Count frequency of each character in the word
        frequency = Counter(word)
        
        # Sort characters based on frequency (most frequent first)
        sorted_chars = sorted(frequency.items(), key=lambda x: x[1], reverse=True)
        
        # Cost array for positions on keypad from 1 to n, where n <= 8 (key '9' can handle more than 3 letters)
        cost = [i for i in range(1, 10)]
        
        # To minimize the total cost:
        # Assign the most frequent letters to the cheapest costs (i.e., 1 push per letter)
        total_pushes = 0
        index = 0
        for char, freq in sorted_chars:
            if index < 8:  # For first 8 characters, we assign them to keys 2 to 9 (one per key)
                total_pushes += cost[index] * freq
                index += 1
            else:  # Remaining characters all go to the last key with cost incrementing by each set of 3
                total_pushes += cost[index] * freq
        
        return total_pushes
